### PR TITLE
Use `--` to separate download name and version.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Use `--` to separate the name and version for cached downloads. Cask download strategies and VCS strategies are already using this, as well as patches.

### CODE REVIEW

**Comment by MikeMcQuaid** [comment]
How will this effect existing entries in the cache?

**Comment by reitermarkus** [comment]
The compatibility layer will rename them if accessed, and `brew cleanup` already uses a regex which also matches the old pattern.

**Comment by MikeMcQuaid** [comment]
@reitermarkus I think it would be good to have some sort of migration which avoids having to redownload anything (particularly SCM checkouts which can be a bit slow).

**Comment by reitermarkus** [comment]
> particularly SCM checkouts which can be a bit slow

The SCM strategies are already using `--` as separator, so these aren't affected.

**Comment by MikeMcQuaid** [comment]
Gotcha 👍. Still think a `brew update` migration might be nice if it's possible.

**Comment by reitermarkus** [comment]
Added the migration to `brew update` instead of `compat`. There is also the `migrate_legacy_cache_if_necessary` method, should we reduce this to just delete the legacy cache? I don't think the files in the legacy cache are relevant anymore.

File: Library/Homebrew/cmd/update-report.rb [hunk 1]
```diff
@@ -183,6 +189,23 @@ def migrate_legacy_cache_if_necessary
...
+    HOMEBREW_CACHE.children.each do |child|
+      next unless child.file?
+
+      next unless /^(?<prefix>[^\.]+[^\-])\-(?<suffix>[^\-].*)/ =~ child.basename.to_s
+      target = HOMEBREW_CACHE/"#{prefix}--#{suffix}"
+
+      if target.exist?
+        FileUtils.rm_r child
+      else
+        FileUtils.mv child, target
```
[the remark this answers is not in the document]
Comment (reitermarkus) [reply Library/Homebrew/cmd/update-report.rb]: Changed it to use the `--force` versions of these.

File: Library/Homebrew/cmd/update-report.rb [hunk 2]
```diff
@@ -183,6 +189,23 @@ def migrate_legacy_cache_if_necessary
...
+    HOMEBREW_CACHE.children.each do |child|
+      next unless child.file?
+
+      next unless /^(?<prefix>[^\.]+[^\-])\-(?<suffix>[^\-].*)/ =~ child.basename.to_s
+      target = HOMEBREW_CACHE/"#{prefix}--#{suffix}"
+
+      if target.exist?
+        FileUtils.rm_rf child
+      else
+        FileUtils.mv child, target, force: true
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/update-report.rb]: I still reckon it's worth handling permission exceptions here a bit more nicely e.g. from below:

```ruby
     rescue Errno::EACCES, Errno::ENOENT
       ofail <<~EOS
         Could not create symlink at #{dst}!
         Please do this manually with:
           sudo ln -sf #{src} #{dst}
           sudo chown $(whoami) #{dst}
       EOS
     end
```

Comment (reitermarkus) [reply Library/Homebrew/cmd/update-report.rb]: Added warnings for permissions exceptions.

**Comment by MikeMcQuaid** [comment]
Last thing: can you have a read of https://github.com/Homebrew/brew/issues/568 before merging and just make sure there's no way that this will or can help with any of the issues described there. Thanks again and 🚢 when ready!

**Comment by ilovezfs** [comment]
The migration seems to have resulted in double `--` when `-` is part of the name. For example, I now have a file
```
reattach--to--user--namespace--2.7.tar.gz
```

But fetching it again with `brew fetch -s reattach-to-user-namespace` results in
```
reattach-to-user-namespace--2.7.tar.gz
```
and the previously cached item is not found.

Also, `bump-formula-pr` seems to be downloading with double hyphen. For example, I just ended up with `firebase--cli--4.0.3.tgz` in my cache when opening https://github.com/Homebrew/homebrew-core/pull/30824

**Comment by ilovezfs** [comment]
I think this may have broken the use of hard links in test-bot https://github.com/Homebrew/homebrew-test-bot/blob/59c69a1bbc8f29daacd8b2d58c851a062a35c88c/cmd/brew-test-bot.rb#L781

The bottle file name hard linked into the cache won't be the same as what you'd get from a download so I don't think there will be a cache hit.

**Pull request merged**